### PR TITLE
QUIC Conformance Updates: Frame, Packet and Stream State Handling

### PR DESCRIPTION
**Pull request by hlandau**
New in this PR:

```
3d0f879f23 QUIC QSM: Get rid of recv_fin_retired in favour of recv_state
093caab4d4 QUIC QSM: Update API documentation
e1413f237f QUIC CONFORMANCE/APL: Handle FIN/reset retirement correctly
948e3d1bc0 QUIC Send Stream State: Transition to DATA_SENT
d7ee7d2080 QUIC CONFORMANCE: Wire the DATA_SENT state
87c6cecdac QUIC APL: Validate receive stream state
6f2b723a5f QUIC APL: Validate send stream state
681f4b36e0 QUIC CONFORMANCE: Stop handling frames after termination
0f4bd9da9b QUIC Conformance: Frame Handling Tests
34cd1746c2 QUIC CONFORMANCE: RFC 9000 s. 9.6
1b8729baa0 QUIC CONFORMANCE: Validate preferred_addr transport parameter
5d72419526 QUIC CONFORMANCE: RFC 9000 s. 19.16: RETIRE_CONNECTION_ID frames
1047e1494e QUIC CONFORMANCE: RFC 9000 s. 19.15: NEW_CONNECTION_ID frames
9fb58f3b1b QUIC RXDP: Make ACK eliciting definition more resilient and centralised
9f5b13bcf8 QUIC CONFORMANCE: RFC 9000 s. 19.14: STREAMS_BLOCKED Frames
35c5e6ac15 QUIC CONFORMANCE: RFC 9000 s. 19.13: STREAM_DATA_BLOCKED Frames
1b65314407 QUIC CONFORMANCE: RFC 9000 s. 19.7
5c58c030cc RFC 9000 s. 19.8: Enforce maximum stream size
51c5087904 QUIC CONFORMANCE: RFC 9000 s. 17.2.5.1
4aed860616 QUIC CONFORMANCE: RFC 9000 s. 17.2.2: Enforce no initial token from server
6b829c0ffc QUIC CONFORMANCE: Enforce packet header reserved bits
a7b36cf8c6 QUIC WIRE: Allow encoding/decoding of reserved header bits
2491024729 QUIC CONFORMANCE: RFC 9000 s. 13.3: MAX_STREAM_DATA generation
3a8e75e2c0 QUIC CONFORMANCE: RFC 9000 s. 12.5: Application CONNECTION_CLOSE frame masking
16d5cd141f QUIC CONFORMANCE: RFC 9000 s. 12.5: Ensure CFQ can not be used to send disallowed frame types in a given PN space
6671f9178e QUIC CONFORMANCE: RFC 9000 s. 12.3: PN Limit
3e75935c8b QUIC CONFORMANCE: RFC 9000 s. 12.3: PN duplicate suppression
ac8b561245 QUIC ACKM: Clarify the role of is_inflight
acca8026db QUIC CONFORMANCE: Enforce minimal frame type encoding
aae6a4a0ce QUIC CONFORMANCE: Packet handling fixes
a472d8a68b QUIC CONFORMANCE: Handle RESET_STREAM final size correctly
f413aaf2a4 QUIC CONFORMANCE: Validate RESET_STREAM final sizes correctly
f0bb62b882 QUIC: Note that we do not retransmit stream data for retransmitted streams
ea4a0f645c QUIC QSM: Free unneeded stream buffers, calculate RESET_STREAM final size correctly
a21a46d893 QUIC QSM: Model final sizes and handle STOP_SENDING correctly
1d21787268 QUIC CONFORMANCE: RFC 9000 s. 3.3: Stream States — Permitted Frame Types — STREAM
a6066c6966 QUIC QSM/STREAM: Refactor to use RFC stream states
```

### CODE REVIEW

**Comment by mattcaswell** [comment]
This looks like it needs a rebase

**Comment by hlandau** [comment]
Updated.

File: ssl/quic/quic_stream_map.c [hunk 1]
```diff
@@ -312,19 +345,251 @@ void ossl_quic_stream_map_update_state(QUIC_STREAM_MAP *qsm, QUIC_STREAM *s)
...
+                                              QUIC_STREAM *qs)
+{
+    switch (qs->send_state) {
+    default:
+        /* Wrong state - caller error. */
+    case QUIC_SSTREAM_STATE_NONE:
+        /* Stream without send part - caller error. */
+        return 0;
+
+    case QUIC_SSTREAM_STATE_SEND:
```
Comment (mattcaswell) [inline ssl/quic/quic_stream_map.c:390]: Might be simpler just to do (same comment for next function):

```` C
if (qs->send_state != QUIC_SSTREAM_STATE_SEND)
    return 0;
````

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I started to review this but only got as far as the first commit :-)

My comments so far...

File: ssl/quic/quic_channel.c [hunk 2]
```diff
@@ -1812,11 +1835,27 @@ static void ch_rx_handle_packet(QUIC_CHANNEL *ch)
...
+     * SHOULD discard these packets."
+     *
+     * We need to be a bit careful here as due to the BIO abstraction layer an
+     * application is liable to be weird and lie to us about peer addresses.
+     * Only apply this check if we actually are using a real address and haven't
+     * been given AF_UNSPEC by the application.
+     */
+    if (!ch->is_server
+        && ch->qrx_pkt->peer != NULL
+        && BIO_ADDR_family(&ch->cur_peer_addr) != AF_UNSPEC
```
Comment (mattcaswell) [inline ssl/quic/quic_channel.c:1854]: Maybe we should only apply the check if family is AF_INET or AF_INET6?

File: test/quic_multistream_test.c [hunk 3]
```diff
@@ -2081,15 +2089,18 @@ static int script_20_inject_plain(struct helper *h, QUIC_PKT_HDR *hdr,
...
+        goto err;
 
     if (!qtest_fault_prepend_frame(h->qtf, frame_buf, written))
-        return 0;
+        goto err;
 
-    return 1;
+    ok = 1;
+err:
+    WPACKET_finish(&wpkt);
```
Comment (mattcaswell) [inline test/quic_multistream_test.c:2102]: Really we should be using WPACKET_cleanup() in the event of error and WPACKET_finish() for successful completion.

Here and below.

File: include/internal/quic_stream_map.h [hunk 4]
```diff
@@ -354,27 +604,173 @@ void ossl_quic_stream_map_update_state(QUIC_STREAM_MAP *qsm, QUIC_STREAM *s);
  */
 void ossl_quic_stream_map_set_rr_stepping(QUIC_STREAM_MAP *qsm, size_t stepping);
 
+
```
Comment (mattcaswell) [inline include/internal/quic_stream_map.h:607]: Spurious additional blank line here

File: ssl/quic/quic_stream_map.c [hunk 5]
```diff
@@ -312,19 +350,269 @@ void ossl_quic_stream_map_update_state(QUIC_STREAM_MAP *qsm, QUIC_STREAM *s)
...
+        qs->send_state = QUIC_SSTREAM_STATE_RESET_RECVD;
+        return 1;
+
+    case QUIC_SSTREAM_STATE_RESET_RECVD:
+        /* Already in the correct state. */
+        return 1;
+    }
+}
+
+/* Stream Receive Part State Management
```
Comment (mattcaswell) [inline ssl/quic/quic_stream_map.c:498]: Wrong multiline comment style

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_stream_map.c (hunk 1 above)
Comment (hlandau) [reply ssl/quic/quic_stream_map.c:390]: I could. I chose to do it like this as I think it makes the state-based handling more explicit/easier to understand.

File: ssl/quic/quic_stream_map.c [hunk 6]
```diff
@@ -435,7 +433,6 @@ int ossl_quic_stream_map_reset_stream_send_part(QUIC_STREAM_MAP *qsm,
 
     case QUIC_SSTREAM_STATE_READY:
     case QUIC_SSTREAM_STATE_SEND:
-    case QUIC_SSTREAM_STATE_DATA_SENT:
         /*
          * If we already have a final size (e.g. because we are coming from
          * DATA_SENT), we have to be consistent with that, so don't change it.
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_stream_map.c:445]: This comment is explaining why `case QUIC_SSTREAM_STATE_DATA_SENT:` appears below bypassing the change to `send_final_size`, so it makes sense as I read it...

**Comment by hlandau** [comment]
Updated.

**Review by mattcaswell — APPROVED** [review]

File: ssl/quic/quic_txp.c [hunk 7]
```diff
@@ -1878,10 +1959,23 @@ static int txp_generate_stream_related(OSSL_QUIC_TX_PACKETISER *txp,
...
         }
 
-        /* Stream Data Frames (STREAM) */
-        if (stream->sstream != NULL) {
+        /*
+         * Stream Data Frames (STREAM)
+         *
+         * RFC 9000 s. 3.3: A sender MUST NOT send a STREAM [...] frame for a
+         * stream in the "Reset Sent" state [or any terminal state]. We don't
+         * send any moore STREAM frames if we are sending, have sent, or are
```
Comment (t8m) [inline ssl/quic/quic_txp.c:1967]: Typo: more

File: test/quic_multistream_test.c [hunk 8]
```diff
@@ -416,6 +437,14 @@ static void helper_cleanup(struct helper *h)
     BIO_free(h->c_net_bio_own);
     h->c_net_bio_own = NULL;
 
+    BIO_free(h->s_qtf_wbio_own);
+    h->s_qtf_wbio_own = NULL;
+
+    if (h->qtf != NULL) {
```
Comment (t8m) [inline test/quic_multistream_test.c:443]: This `if` is unnecessary.

**Review by t8m — CHANGES_REQUESTED** [review]
Just a few minor nits

File: ssl/quic/quic_rx_depack.c [hunk 9]
```diff
@@ -495,6 +569,16 @@ static int depack_do_frame_stream(PACKET *pkt, QUIC_CHANNEL *ch,
                                       frame_data.is_fin))
         return 0;
 
+    /*
+     * rs_fin will be 1 only if we can read all data up to and including the FIN
+     * without any gaps before it; this implies we have received all data.
+     */
+    if (!ossl_quic_rstream_available(stream->rstream, &rs_avail, &rs_fin))
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_rx_depack.c:576]: Good idea.

**Comment by hlandau** [comment]
Updated.

MacOS is failing, will look into it.

File: ssl/quic/quic_tserver.c [hunk 10]
```diff
@@ -489,3 +489,15 @@ int ossl_quic_tserver_shutdown(QUIC_TSERVER *srv)
 
     return ossl_quic_channel_is_terminated(srv->ch);
 }
+
+int ossl_quic_tserver_ping(QUIC_TSERVER *srv)
+{
+   if (ossl_quic_channel_is_terminated(srv->ch))
+     return 0;
```
Comment (t8m) [inline ssl/quic/quic_tserver.c:496]: Nit: wrong indentation here.

**Review by t8m — CHANGES_REQUESTED** [review]
One minor nit

**Comment by hlandau** [comment]
CI passed. Updated.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@mattcaswell please reconfirm

**Comment by hlandau** [comment]
Currently failing to repro the macOS failure on @levitte's M1.

**Comment by t8m** [comment]
> Currently failing to repro the macOS failure on @levitte's M1.

IMO this is some intermittent problem, we can defer its solution for now.

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged!

**Pull request closed without merging**